### ISSUE TITLE
How to wait for multiple events to finish?

### ISSUE THREAD

**Issue by jonny166** [circuits/circuits#182]
### Expected Behaviour
If you have a handler that fires multiple events, and you want to wait for all of those events to complete, what is the right way to do this?  The "yield self.wait(e)" only works for the first event and will sit forever after this first yield statement in a handler. 

### Steps to reproduce

```
from __future__ import print_function
from circuits import Event, Component, Worker, task
from circuits import handler, Debugger
from time import sleep

def do_thing(nbr):
    # Some slow task
    sleep(2)
    return "finished"

class App(Component):
    
    def started(self, *args, **kwargs):
        self.worker = Worker(process=2, workers=5)
        self.worker.register(self)
        
        events = []
        for i in range(3):
            e = task(do_thing, i)
            events.append(e)
            self.fire(e, self.worker)

        for e in events:
            print("waiting for %s to finish" % str(e))
            yield self.wait(e)
            print("%s finished" % str(e))

# Start and "run" the system.
app = App()
Debugger().register(app)
app.run()
```

```
<registered[*] (<Debugger/* 83455:MainThread (queued=0) [S]>, <App/* 83455:MainThread (queued=2) [R]> )>
<started[*] (<App/* 83455:MainThread (queued=1) [R]> )>
waiting for <task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 0 )> to finish
<registered[worker] (<Worker/worker 83455:MainThread (queued=0) [S]>, <App/* 83455:MainThread (queued=4) [R]> )>
<task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 0 )>
<task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 1 )>
<task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 2 )>
<task_success[<Worker/worker 83455:MainThread (queued=0) [S]>] (<task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 2 )>, 'finished' )>
<task_success[<Worker/worker 83455:MainThread (queued=0) [S]>] (<task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 1 )>, 'finished' )>
<task_done[<Worker/worker 83455:MainThread (queued=0) [S]>] ('finished' )>
<task_success[<Worker/worker 83455:MainThread (queued=0) [S]>] (<task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 0 )>, 'finished' )>
<task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 0 )> finished
waiting for <task[<Worker/worker 83455:MainThread (queued=0) [S]>] (<function do_thing at 0x1080a7668>, 1 )> to finish
```

### Environment and Platform

- OS:OS X
- Python: 2.7
- circuits: 3.2

----

**Comment by prologic** [comment]
Hi Kathy!

Sorry for the late response (been very busy at work).

circuits API doesn't currently support this out-of-the-box; however how
would you expect the API to look like and behave? Would you be interested
in cutting a PR to add this feature? If so I'd be more than happy to review!

/cc @spaceonc

cheers
James


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by jonny166** [comment]
Hi James, thanks for the response.
I'd guess I'd like to see the API to support the example I gave above, allowing you to call wait more than once in a handler.  Also, a "wait_all" function that takes a list of events would be useful.
I don't really have the time to create a PR for this at the moment, but maybe in the future.

For now maybe just a note in the documentation would be helpful.

Thanks!
Kathy

**Comment by prologic** [comment]
Hi Kathy!

I'll see if I can make some time this weekend to add this -- Are you
willing to test this if I can get a PR ready?

cheers
James


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by spaceone** [comment]
Do you think https://github.com/spaceone/circuits/commit/f8313aabf4ef96c87b0c3736a18ef4e84b23dc96 is fine? This extends the API to allow to pass a list/tuple/iterator of events to component.wait(). It yields all values of all events - I don't know if this is okay.

**Comment by prologic** [comment]
Yeah this looks good. Send a PR let's review it :)


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by apollo13** [comment]
@spaceone Your proposed solution serializes the event dispatching, wouldn't it be nicer to fire them all and then just wait for the results? This way the event handlers could run in parallel (depending on how the code looks like of course)

**Comment by spaceone** [comment]
@apollo13 Yes, of course :D My commit is obviously wrong :D

**Comment by spaceone** [comment]
Hm, they are already fired.

**Comment by apollo13** [comment]
Ok, will dig depper into how event handling works -- still learning ;)

**Comment by prologic** [comment]
@spaceone I think your solution is correct. Can you submit a PR with an associated test?

**Comment by spaceone** [comment]
@prologic @ke4roh 
There is also the problem that the self.wait() will block forever if the second event is already finished.
So we should fix this here, too.
@ke4roh Can provide an example test case for this.

**Comment by spaceone** [comment]
Or maybe in another explicit issue.

**Comment by prologic** [comment]
File a separate issue :)


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by spaceone** [comment]
https://github.com/circuits/circuits/issues/226
